indexer/sql/src/non_blocking_indexer.rs

```diff
@@ -0,0 +1,343 @@
+    }
+}
+
+impl Indexer for NonBlockingIndexer {
+    type Error = crate::error::IndexerError;
+
+    fn start(&mut self) -> error::Result<()> {
+        let context = self.context.clone();
+        block_call(self.runtime.as_ref(), &self.handle, async move {
+            context.migrate_db().await
+        })?;
+
+        self.indexing = true;
+        Ok(())
+    }
+
+    fn shutdown(&mut self) -> error::Result<()> {
+        // Avoid running this twice when called manually and from `Drop`
+        if !self.indexing {
+            return Ok(());
+        }
+        self.indexing = false;
+
+        // NOTE: This is to allow the indexer to commit all db transactions since this is done
+        // async. It just loops quickly making sure no indexing blocks are remaining.
+        for _ in 0..10 {
+            let blocks = self.blocks.lock().expect("Can't lock blocks");
+            if blocks.is_empty() {
+                break;
+            }
+
+            drop(blocks);
+
+            sleep(Duration::from_millis(10));
+        }
+
+        #[cfg(feature = "tracing")]
+        {
+            let blocks = self.blocks.lock().expect("Can't lock blocks");
+            if !blocks.is_empty() {
+                tracing::warn!("Some blocks are still being indexed, maybe non_blocking_indexer `post_indexing` wasn't called by the main app?");
+            }
+        }
+
+        Ok(())
+    }
+
+    fn pre_indexing(&self, _block_height: u64) -> error::Result<()> {
+        assert!(self.indexing, "Can't index after shutdown");
+        Ok(())
+    }
+
+    fn index_block(&self, block: &BlockInfo, _block_outcome: &BlockOutcome) -> error::Result<()> {
+        assert!(self.indexing, "Can't index after shutdown");
+
+        #[cfg(feature = "tracing")]
+        tracing::info!(block_height = block.height, "index_block called");
+        self.find_or_create(block, |_block_to_index| {
+            #[cfg(feature = "tracing")]
+            tracing::info!(block_height = block.height, "index_block started/finished");
+            Ok(())
+        })
+    }
+
+    fn index_transaction(
+        &self,
+        block: &BlockInfo,
+        tx: &Tx,
+        tx_outcome: &TxOutcome,
+    ) -> error::Result<()> {
+        assert!(self.indexing, "Can't index after shutdown");
+
+        #[cfg(feature = "tracing")]
+        tracing::info!(block_height = block.height, "index_transaction called");
+
+        self.find_or_create(block, |block_to_index| {
+            #[cfg(feature = "tracing")]
+            tracing::info!(block_height = block.height, "index_transaction started");
+
+            block_to_index.txs.push((tx.clone(), tx_outcome.clone()));
+
+            #[cfg(feature = "tracing")]
+            tracing::info!(block_height = block.height, "index_transaction finished");
+            Ok(())
+        })
+    }
+
+    fn post_indexing(&self, block_height: u64) -> error::Result<()> {
+        assert!(self.indexing, "Can't index after shutdown");
+
```

Comment:
If this function return a Result, is better to don't panic here. Returning a result is better cause it could be handled